README.md
libndofdev
==========

[![CI](https://github.com/janoc/libndofdev/actions/workflows/CI.yml/badge.svg)](https://github.com/janoc/libndofdev/actions/workflows/CI.yml)

Linux support for the 3Dconnexion SpaceBall, SpaceNavigator and joysticks in the SecondLife clients

Currently supported are the following devices (VID:PID):

  * 0x046d:0xc603   3Dconnexion SpaceMouse (untested)
  * 0x046d:0xc605   CADMan (untested)
  * 0x046d:0xc606   SpaceMouse Classic (untested)
  * 0x046d:0xc621   SpaceBall 5000 (tested, works)
  * 0x046d:0xc623   SpaceTraveller (untested)
  * 0x046d:0xc625   SpacePilot (untested)
  * 0x046d:0xc626   SpaceNavigator (tested, works)
  * 0x046d:0xc627   SpaceExplorer (untested - from Armin Weatherwax <Armin.Weatherwax (at) gmail.com>)
  * 0x046d:0xc628   SpaceNavigator for Notebooks (untested)
  * 0x046d:0xc629   SpacePilot Pro (untested)
  * 0x046d:0xc62b   SpaceMouse Pro (untested - from David Seikel < http://onefang.net/ >

  * 0x256f:0xc62e   SpaceMouse Wireless (cable) (untested)
  * 0x256f:0xc62f   SpaceMouse Wireless (receiver) (untested)
  * 0x256f:0xc631   SpaceMouse Wireless (untested)
  * 0x256f:0xc632   SpaceMouse Pro Wireless (untested)
  * 0x256f:0xc633   SpaceMouse Enterprise (tested, works)
  * 0x256f:0xc635   SpaceMouse Compact (untested - from Kai Werthwein < https://github.com/KaiWerthwein >

The following devices were contributed by Ricky Curtice (http://rwcproductions.com/):
(https://bitbucket.org/lindenlab/3p-libndofdev-linux/pull-request/1/open-277-linux-support-for-spacemousepro/diff)

CADMan, SpaceM (classic & wireless) except SpaceMouse Pro
(non-wireless) and SpaceMouse Entreprise, SpaceNavigator for Notebooks

The devices marked as "untested" are expected to work and their
VID:PIDs were sent to me by external contributors. I haven't tested
them personally because I don't have that hardware, they could work,
could be completely broken or anything in-between. Caveat
emptor. However, please do report any working or not working devices
from the list above - that's the only way to get the problems fixed.

In addition, the library supports every joystick/gamepad device
supported by the SDL library. SDL2 is supported by passing a flag to
the makefile: `USE_SDL2=1 make`


C A U T I O N !!!
-----------------

DO NOT USE THIS LIBRARY FOR ADDING SUPPORT FOR THE 3Dconnexion HARDWARE TO NEW APPLICATIONS!

This library is a messy hack emulating the proprietary 3Dconnexion SDK, in order to simplify the SecondLife
client integration. If you are looking to integrate support for these devices into a new application,
a much better solution is to use the event interface directly. An example of how to do this can be found
here:
  * http://jciger.com/files/software/linux/spacenav/spacenavig.c

If you are building a virtual reality application, then the best option is to use the VRPN library:
  * http://www.cs.unc.edu/Research/vrpn/

VRPN supports all the devices above and many more in a transparent and abstract manner, so you don't have
to add support for each different interaction device or tracker
specially.

Another option simulating the official 3DConnexion driver is the
Spacenav project:
   * https://spacenav.sourceforge.net/


CHANGELOG
2023-12-13  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.14
            - Support for SpaceMouse Enterprise
	    - Fixed crashes if device reports more than 32 buttons/6 axes

2021-04-02  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.13
            - Support for joystick hats as extra axes (thanks @zi-ree)

2019-12-25  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.12
            - Support for SDL2 (thanks @Nicky-D)

2019-11-01  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.11
            - Updated API to match the new libndof library version used by
	the Linden Lab

2018-03-23  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.10
            - better version number handling so that there is no need to parse Changelog entries for it anymore

2018-11-06  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.9
            - added support for SpaceMouse Compact (thanks Kai Werthwein)

2015-10-11  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.8
            - improved detection of number of axes & buttons (thanks @techwolf)
            - added a simple test program to verify basic functionality

2015-10-09  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.7
            - fixed dependency on libexplain that sneaked in in the last version (thanks @techwolf)
            - fixed two ueber-strict warnings about return values from write() (thanks @techwolf)

2014-10-20  Jan Ciger <jan.ciger (at) gmail.com>
        * 0.6
            - merged patch from Ricky Curtice (http://rwcproductions.com/)
                - untested support for more devices
                - dynamic determination of the number of axes
                - cleanups

2014-04-12  Jan Ciger  <jan.ciger (at) gmail.com>
        * 0.5
            - added support for 3Dconnexion SpaceMouse Pro & increased number of supported buttons to 32 (thanks to David Seikel)
2014-04-08  Jan Ciger  <jan.ciger (at) gmail.com>
        * 0.4
            - added support for 3Dconnexion SpaceBall 5000 & increased number of supported buttons to 12
            - merged change from Armin Weatherwax <Armin.Weatherwax (at) gmail.com> for 3Dconnexion SpaceExplorer
2010-08-10  Jan Ciger  <jan.ciger (at) gmail.com>
        * 0.3
            - fixed support for newer kernels - it seems that somewhere between 2.6.31 and 2.6.35 the SpaceNavigator
            driver was changed to send EV_ABS events instead of EV_REL. Thanks to Thomax G. thomax23 (at) googlemail.com

2008-08-25  Jan Ciger  <jan.ciger (at) gmail.com>

        * 0.2
            - fixed button support (Do'h!)
            - turn the LED on when the SpaceNavigator is opened and off when it is closed

2008-08-20  Jan Ciger  <jan.ciger (at) gmail.com>

        * 0.1 Initial release


ndofdev_test.c
/*
*
* Copyright (c) 2008, Jan Ciger (jan.ciger (at) gmail.com)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*     * Redistributions of source code must retain the above copyright
*       notice, this list of conditions and the following disclaimer.
*     * Redistributions in binary form must reproduce the above copyright
*       notice, this list of conditions and the following disclaimer in the
*       documentation and/or other materials provided with the distribution.
*     * The name of its contributors may not be used to endorse or promote products
*       derived from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY Jan Ciger ''AS IS'' AND ANY
* EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL Jan Ciger BE LIABLE FOR ANY
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
 * Basic sanity test that the library works and can actually read device data.
 */

#include <stdio.h>
#include <unistd.h>

#include "ndofdev_external.h"

int main(int argc, char** argv)
{
    int i;
    NDOF_Device *dev = NULL;

    ndof_libinit(NULL, NULL, NULL);

    dev = ndof_create();
    if(ndof_init_first(dev, NULL) < 0)
    {
        perror("Unable to open any NDOF devices!");
        return(-1);
    }

    puts("Detected:");
    ndof_dump(stdout, dev);

    puts("\nPress RETURN to start polling ...");
    getchar();

    // try to poll the device and get data
    while(1)
    {
        ndof_update(dev);

        printf("\nAxes: ");
        for(i = 0; i < dev->axes_count; ++i)
            printf("%ld ", dev->axes[i]);

        printf("\nButtons: ");
        for(i = 0; i < dev->btn_count; ++i)
            printf("%ld ", dev->buttons[i]);

        usleep(100000);
    }
}


ndofdev_external.h
/*
 @file ndofdev_external.h
 @brief libndofdev main header file for external use.
 
 Copyright (c) 2007, 3Dconnexion, Inc. - All rights reserved.
 
 Redistribution and use in source and binary forms, with or without 
 modification, are permitted provided that the following conditions are met:
 - Redistributions of source code must retain the above copyright notice, 
   this list of conditions and the following disclaimer.
 - Redistributions in binary form must reproduce the above copyright notice, 
   this list of conditions and the following disclaimer in the documentation 
   and/or other materials provided with the distribution.
 - Neither the name of the 3Dconnexion, Inc. nor the names of its contributors 
   may be used to endorse or promote products derived from this software 
   without specific prior written permission.
 
 THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE 
 DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR 
 ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
 (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; 
 LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON 
 ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT 
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef __ndofdev_external_h__
#define __ndofdev_external_h__

#ifdef __cplusplus
extern "C" {
#endif

#define NDOF_MAX_AXES_COUNT         8
#define NDOF_MAX_BUTTONS_COUNT      32

typedef enum NDOF_HotPlugResult {
    NDOF_KEEP_HOTPLUGGED,
    NDOF_DISCARD_HOTPLUGGED
} NDOF_HotPlugResult;

/** Do NOT create NDOF_Device variables manually. Always use ndof_create. */
typedef struct NDOF_Device {
    long axes[NDOF_MAX_AXES_COUNT];           /* axes current values */
    long buttons[NDOF_MAX_BUTTONS_COUNT];     /* buttons current values */
    short axes_count;		/* actual # axes (self sensed on OS X) */
    short btn_count;		/* actual # buttons (self sensed on OS X) */
    long axes_max;			/* logical max */
    long axes_min;			/* logical min */
	unsigned char absolute;	/* if 0 calc delta. On OSX we always have deltas. */
	unsigned char valid;	/* if 0, clients should not access this device. */
	char manufacturer[256]; /* name of device manufacturer */
    char product[256];      /* name of the device */
    void *private_data;     /* ptr to platform specific/private data */
} NDOF_Device;

/** Callback type for new hot-plugged devices. 
 *  Parameters: dev - pointer to the newly added NDOF device. A new NDOF_Device
 *                    struct is allocated for the client. Clients can express
 *                    interest in the device by returning NDOF_KEEP_HOTPLUGGED,
 *                    otherwise the memory pointed by dev will be released. */
typedef NDOF_HotPlugResult (*NDOF_DeviceAddCallback)(NDOF_Device *dev);

/** Removal callback type. Parameter points to the removed device.  */
typedef void (*NDOF_DeviceRemovalCallback)(NDOF_Device *device);

/** Purpose:    Initializes the library. 
 *  Parameters: in_add_cb - Callback invoked when a new	device is hotplugged in.
 *                          Pass NULL if you don't care.
 *              in_removal_cb - Callback invoked when a device is removed.
 *                              Pass NULL if you don't care.
 *              platform_specific - On Windows, it can be a pointer to a already
 *                                  initialized LPDIRECTINPUT8 value; pass NULL 
 *                                  for full initialization. Unused on OS X.
 *  Notes:      The callbacks functionality is currently implemented on Mac OS X
 *              only.  The callbacks functions are currently ignored on Windows.
 *  Returns:    0 if ok. 
 */
extern int ndof_libinit(NDOF_DeviceAddCallback in_add_cb, 
                        NDOF_DeviceRemovalCallback in_removal_cb,
                        void *platform_specific);

/** Purpose:    Clean up.  Must be called before program termination. 
 */
extern void ndof_libcleanup();

/** Purpose:    Allocates memory for a new NDOF_Device struct and initializes 
 *              it with default values. 
 *  Notes:      Always use this function to allocate a new NDOF_Device. 
 *              Call ndof_libcleanup to free the memory. 
 *  Returns:    Pointer to a mallocated struct.
 */
extern NDOF_Device *ndof_create();

/** Purpose:    Get the _first_ NDOF device found in the USB bus. 
 *  Notes:      It will be attempted to use the max and min values of the 
 *              input struct as the range of returned values from the device. 
 *              This may not work for every device. In any case, when this 
 *              function returns, the axes_min, axes_max values will contain 
 *              the actual range.
 *  Parameters: in_out_dev - Must be allocated externally using ndof_create(). 
 *              param - On Windows, you may use this to pass in a 
 *                      LPDIRECTINPUTDEVICE8.  It will be attempted to use
 *                      that instead of creating a new one. In all other cases
 *                      just pass NULL.
 *  Returns:    0 if all is ok, -1 otherwise. 
 */
extern int ndof_init_first(NDOF_Device *in_out_dev, void *param);

/** Purpose:    Reads the current status of the input device. 
 *  Parameters: Must be initialized with ndof_create(). 
 */
extern void ndof_update(NDOF_Device *in_dev);

/** Purpose:    Dumps device info on specified FILE*. */
extern void ndof_dump(FILE* stream, NDOF_Device *dev);

/** Purpose:    Dumps list of NDOF devices currently in use on specified FILE*. */
extern void ndof_dump_list(FILE* stream);

#if TARGET_OS_MAC
/** Returns the number of connected NDOF devices. Implemented only on OS X. */
extern int ndof_devcount();
#endif

#ifdef __cplusplus
}
#endif
    
#endif /* __ndofdev_external_h__ */


ndofdev.c
/*
*
* Copyright (c) 2008, Jan Ciger (jan.ciger (at) gmail.com)
* All rights reserved.
*
* Redistribution and use in source and binary forms, with or without
* modification, are permitted provided that the following conditions are met:
*     * Redistributions of source code must retain the above copyright
*       notice, this list of conditions and the following disclaimer.
*     * Redistributions in binary form must reproduce the above copyright
*       notice, this list of conditions and the following disclaimer in the
*       documentation and/or other materials provided with the distribution.
*     * The name of its contributors may not be used to endorse or promote products
*       derived from this software without specific prior written permission.
*
* THIS SOFTWARE IS PROVIDED BY Jan Ciger ''AS IS'' AND ANY
* EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
* WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
* DISCLAIMED. IN NO EVENT SHALL Jan Ciger BE LIABLE FOR ANY
* DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
* (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
* ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
* (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
* SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

/*
    * Drop-in replacement for the 3D Connexion SDK library not available on Linux
    *
    * Uses native Linux input (evdev) device for SpaceNavigator, otherwise SDL for
    * regular joysticks.
    *
    * Evdev could be used for joysticks as well, but higher level logic would have to be
    * re-implemented (calibration, filtering, etc.) - SDL includes it already.
    *
    * Release 0.9
*/

#include <linux/input.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#ifdef USE_SDL2
#include <SDL2/SDL.h>
#else
#include <SDL/SDL.h>
#endif


#include "ndofdev_external.h"
#include "ndofdev_version.h"

// Hack - we need to store the descriptor
// of the SpaceNavigator in order to be able to
// turn its LED off in the ndof_cleanup() function :(
static int spacenav_fd = -1;

// SpaceMouse button remapping table - SpaceMouse Enterprise has only
// 32 buttons but reports 9bit button codes, likely for compatibility
// with old hw.
#define SPACE_MOUSE_BUTTON_SPACE 256
static long SPACE_MOUSE_BUTTON_MAPPING[SPACE_MOUSE_BUTTON_SPACE];

const int SPACE_NAVIG_THRESHOLD = 20; // minimum change threshold

typedef struct
{
    // native HID interface or SDL?
    int USE_SDL;

    // SpaceNavigator
    int fd;             // file descriptor of the device
    long int *axes;     // last state
    long int *buttons;

    // SDL joysticks
    SDL_Joystick *j;

    int num_hats;       // how many hats are on this joystick

} LinJoystickPrivate;

NDOF_Device *ndof_create()
{
    NDOF_Device *d = (NDOF_Device *) calloc(1, sizeof(NDOF_Device));
    return d;
}

#define test_bit(nr, addr) \
     (((1UL << ((nr) % (sizeof(long) * 8))) & ((addr)[(nr) / (sizeof(long) * 8)])) != 0)
#define NBITS(x) ((((x)-1)/(sizeof(long) * 8))+1)

int ndof_init_first(NDOF_Device *in_out_dev, void *param)
{
    // try to find 3DConnexion SpaceNavigator first
    int i = 0;
    int fd = -1;
    char *fname = NULL;
    struct input_id ID;

    printf("libndofdef version %d.%d, (c) Jan Ciger https://github.com/janoc/libndofdev\n", NDOFDEV_MAJOR, NDOFDEV_MINOR);

    fname = (char *)malloc(1000 * sizeof(char));
    while (i < 32)
    {
        sprintf(fname, "/dev/input/event%d", i++);
        fd = open(fname, O_RDWR | O_NONBLOCK);
        if (fd > 0)
        {
            ioctl(fd, EVIOCGID, &ID);        // get device ID
            if (                             // For a nice list see http://spacemice.org/index.php?title=Dev
                ((ID.vendor == 0x046d) &&    // Logitech's Vendor ID, used by 3DConnexion until they got their own.
                    (
                        (ID.product == 0xc603) || // SpaceMouse (untested)
                        (ID.product == 0xc605) || // CADMan (untested)
                        (ID.product == 0xc606) || // SpaceMouse Classic (untested)
                        (ID.product == 0xc621) || // SpaceBall 5000
                        (ID.product == 0xc623) || // SpaceTraveler (untested)
                        (ID.product == 0xc625) || // SpacePilot (untested)
                        (ID.product == 0xc626) || // SpaceNavigators
                        (ID.product == 0xc627) || // SpaceExplorer (untested)
                        (ID.product == 0xc628) || // SpaceNavigator for Notebooks (untested)
                        (ID.product == 0xc629) || // SpacePilot Pro (untested)
                        (ID.product == 0xc62b) || // SpaceMousePro
                        0
                    )
                ) ||
                ((ID.vendor == 0x256F) &&    // 3Dconnexion's Vendor ID
                    (
                        (ID.product == 0xc62E) || // SpaceMouse Wireless (cable) (untested)
                        (ID.product == 0xc62F) || // SpaceMouse Wireless (receiver) (untested)
                        (ID.product == 0xc631) || // Spacemouse Wireless (untested)
                        (ID.product == 0xc632) || // Spacemouse Pro Wireless (untested)
                        (ID.product == 0xc633) || // Spacemouse Enterprise
                        (ID.product == 0xc635) || // Spacemouse Compact (untested)
                        0
                    )
                ))
            {
                printf("Using evdev device: %s\n", fname);
                break;
            } else {
                close(fd);
                fd = -1;
            }
        }
    }

    if(fd > 0)
    {
        // We have SpaceNavigator, use it
        spacenav_fd = fd;

        // default to sane values for these devices
        unsigned int axes_count = 6;
        unsigned int button_count = 32;

        // Get the actual number of axes for this device.
        int detected_axes_count = 0;

        // first absolute axes
        unsigned long absbit[NBITS(ABS_MAX)] = { 0 };
        if (ioctl(fd, EVIOCGBIT(EV_ABS, sizeof(absbit)), absbit) >= 0)
        {
            for (i = 0; i < ABS_MISC; ++i)
            {
                /* Skip hats */
                if (i == ABS_HAT0X)
                {
                    i = ABS_HAT3Y;
                    continue;
                }

                if (test_bit(i, absbit))
                    detected_axes_count++;
            }
        } else {
            perror("Failed to obtain the number of absolute axes for device:\n");
        }

        // now relative axes
        unsigned long relbit[NBITS(REL_MAX)] = { 0 };
        if (ioctl(fd, EVIOCGBIT(EV_REL, sizeof(relbit)), relbit) >= 0)
        {
            for (i = 0; i < REL_MISC; ++i)
            {
                if (test_bit(i, relbit))
                    detected_axes_count++;
            }
        } else {
            perror("Failed to obtain the number of relative axes for device:\n");
        }

        if (detected_axes_count != 0)
            axes_count = detected_axes_count;

        // Get the actual number of buttons for this device.
        int detected_button_count = 0;
        unsigned long keybit[NBITS(KEY_MAX)] = { 0 };
        if (ioctl(fd, EVIOCGBIT(EV_KEY, sizeof(keybit)), keybit) >= 0)
        {
            for (i = BTN_JOYSTICK; i < KEY_MAX; ++i)
            {
                if (test_bit(i, keybit))
                    detected_button_count++;
            }

            for (i = BTN_MISC; i < BTN_JOYSTICK; ++i)
            {
                if (test_bit(i, keybit))
                    detected_button_count++;
            }

            if (detected_button_count != 0)
                button_count = detected_button_count;
        } else {
            perror("Failed to obtain the number of buttons for device:\n");
        }

        // clamp number of axes & buttons to what is actually
        // supported by the original SDK to avoid corrupting
        // memory. 3DConnexion devices routinely report 255+ buttons
        // even though they have only 32.
        in_out_dev->axes_count = (axes_count <= NDOF_MAX_AXES_COUNT) ? axes_count : NDOF_MAX_AXES_COUNT;
        in_out_dev->btn_count  = (button_count <= NDOF_MAX_BUTTONS_COUNT) ? button_count : NDOF_MAX_BUTTONS_COUNT;
        in_out_dev->absolute   = 0;
        in_out_dev->valid      = 1;
        in_out_dev->axes_max   = 512;
        in_out_dev->axes_min   = -512;
        ioctl(fd, EVIOCGNAME(255), &in_out_dev->product); // copy name of the device

        // private data
        LinJoystickPrivate *priv = (LinJoystickPrivate *) malloc (sizeof(LinJoystickPrivate));
        priv->fd = fd;
        priv->axes = (long int *) calloc(axes_count, sizeof(long int));
        priv->buttons = (long int *) calloc(button_count, sizeof(long int));
        priv->USE_SDL = 0;
        priv->j = NULL;
        in_out_dev->private_data = priv;

        // turn the LED on
        struct input_event led_ev;

        led_ev.type = EV_LED;
        led_ev.code = LED_MISC;
        led_ev.value = 1;
        if(write(spacenav_fd, &led_ev, sizeof(struct input_event)) < 0)
            perror("Failed to write LED_ON command:\n");

        return 0;

    } else {
        // SpaceNavigator not found, use SDL Joystick
        SDL_Joystick *j = SDL_JoystickOpen(0);
        if(j)
        {
            in_out_dev->axes_count = SDL_JoystickNumAxes(j) + SDL_JoystickNumHats(j) * 2; // each hat has 2 axes
            in_out_dev->btn_count = SDL_JoystickNumButtons(j);
            in_out_dev->absolute = 0; // always relative on Linux
            in_out_dev->valid = 1;
            in_out_dev->axes_max = 32767;
            in_out_dev->axes_min = -32767;
#ifdef USE_SDL2
            strncpy(in_out_dev->product, SDL_JoystickName(j), 255);
#else
            strncpy(in_out_dev->product, SDL_JoystickName(0), 255);
#endif
            // private data
            LinJoystickPrivate *priv = (LinJoystickPrivate *) malloc (sizeof(LinJoystickPrivate));
            priv->j = j;
            priv->fd = -1;
            priv->USE_SDL = 1;
            // remember number of hats for later axis mapping in ndof_update()
            priv->num_hats = SDL_JoystickNumHats(j);
            in_out_dev->private_data = priv;

            return 0;
        } else {

            // no joysticks found
            return -1;
        }
    }
}

void ndof_libcleanup()
{
    if(spacenav_fd > 0)
    {
        // turn the LED off
        struct input_event led_ev;

        led_ev.type = EV_LED;
        led_ev.code = LED_MISC;
        led_ev.value = 0;
        if(write(spacenav_fd, &led_ev, sizeof(struct input_event)) < 0)
            perror("Failed to write LED_OFF command:\n");
    }

    // FIXME: needs to cleanup the memory
}

int ndof_libinit(NDOF_DeviceAddCallback in_add_cb,
                 NDOF_DeviceRemovalCallback in_removal_cb,
                 void *platform_specific)
{
    // Fill the SpaceMouse button mapping table
    //
    // Remap the buttons with high number codes into the unused spaces
    // within the first 32 buttons - likely not how the Windows driver
    // does it but at least they will be usable. We also ignore the
    // 9th bit.

    for(long i = 0; i < SPACE_MOUSE_BUTTON_SPACE; ++i)
        SPACE_MOUSE_BUTTON_MAPPING[i] = i;

    SPACE_MOUSE_BUTTON_MAPPING[282 & 0xff] = 3;  /* Rotation lock */
    SPACE_MOUSE_BUTTON_MAPPING[291 & 0xff] = 6;  /* Enter */
    SPACE_MOUSE_BUTTON_MAPPING[292 & 0xff] = 7;  /* Delete */
    SPACE_MOUSE_BUTTON_MAPPING[430 & 0xff] = 9;  /* Tab */
    SPACE_MOUSE_BUTTON_MAPPING[431 & 0xff] = 26; /* Space */
    SPACE_MOUSE_BUTTON_MAPPING[358 & 0xff] = 27; /* V1 */
    SPACE_MOUSE_BUTTON_MAPPING[359 & 0xff] = 28; /* V2 */
    SPACE_MOUSE_BUTTON_MAPPING[360 & 0xff] = 29; /* V3 */
    SPACE_MOUSE_BUTTON_MAPPING[332 & 0xff] = 30; /* 11 */
    SPACE_MOUSE_BUTTON_MAPPING[333 & 0xff] = 31; /* 12 */

    // Initialize the joystick subsystem
    SDL_InitSubSystem(SDL_INIT_JOYSTICK);

    return 0;
}

void ndof_update(NDOF_Device *in_dev)
{
    int i;

    LinJoystickPrivate *priv = (LinJoystickPrivate *) in_dev->private_data;
    assert(priv != NULL);

    if(priv->USE_SDL)
    {
        SDL_JoystickUpdate();
        SDL_Joystick *j = priv->j;

        for(i = 0; i < in_dev->axes_count - priv->num_hats*2; i++) // hats will get mapped to uppermost axes
        {
            in_dev->axes[i] = (int) (SDL_JoystickGetAxis(j, i));
        }

        for(i = 0; i < priv->num_hats; i++)
        {
            int x = 0;
            int y = 0;

            int value = SDL_JoystickGetHat(j, i);

            // map hat values (1, 2, 4, 8) to axis data (-32768, 32768)
            if(value & 1)
            {
                y = -32767;
            }
            else if(value & 4)
            {
                y = 32768;
            }
            if(value & 2)
            {
                x = 32767;
            }
            else if(value & 8)
            {
                x = -32768;
            }

            // add the hat data to the uppermost axes data
            in_dev->axes[in_dev->axes_count - priv->num_hats*2 + i*2  ] = x;
            in_dev->axes[in_dev->axes_count - priv->num_hats*2 + i*2+1] = y;
        }

        for(i = 0; i < in_dev->btn_count; i++)
        {
            in_dev->buttons[i] = SDL_JoystickGetButton(j, i);
        }
    } else {
        // update SpaceNavigator

        struct input_event ev;

        while(read(priv->fd, &ev, sizeof(struct input_event)) > 0)
        {
            switch (ev.type)
            {
                case EV_KEY:
                    int mapped_code = SPACE_MOUSE_BUTTON_MAPPING[ev.code & 0xff];
                    // printf("Key %d (mapped to %d) pressed %d .\n", ev.code, mapped_code, ev.value);
                    priv->buttons[mapped_code] = ev.value;
                    break;

                case EV_REL:
                case EV_ABS: // 2.6.35 and up, maybe earlier kernels too send EV_ABS instead of EV_REL
                    // printf("%d %g\n", ev.code, ev.value);

                    // clean up small values
                    priv->axes[ev.code] = abs(ev.value) > SPACE_NAVIG_THRESHOLD ? ev.value : 0;
                    break;

                default:
                    break;
            }
        }

        memcpy(in_dev->axes, priv->axes, in_dev->axes_count * sizeof(long int));
        memcpy(in_dev->buttons, priv->buttons, in_dev->btn_count * sizeof(long int));
    }
}

void ndof_dump(FILE *stream, NDOF_Device *dev)
{
    fprintf(stream, "NDOF device: %s, with %d buttons and %d axes\n", dev->product, dev->btn_count, dev->axes_count);
}

void ndof_dump_list(FILE *stream)
{

}
